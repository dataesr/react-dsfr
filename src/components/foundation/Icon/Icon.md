## Icons - Icônes

Toutes les icônes de la librairie [remixicon](https://remixicon.com/) sont utilisables.

### Usage simple
```jsx
import Icon from './Icon';

<Icon
    name="ri-bubble-chart-line"
    size="1x">
    <p>Lorem ipsum</p>
</Icon>
```

```jsx
import Icon from './Icon';
<Icon
    className="ds-fr--v-middle"
    name="ri-music-2-line"
    iconPosition="right"
    size="lg">
    <h3>Title</h3>
</Icon>
```

### Usage avec un composant Button

```jsx
import Icon from './Icon';
import Button from '../../interface/Button';

<Button
    title="title"
    icon="ri-calendar-todo-fill"
    iconPosition="right">click
</Button>
```

### Usage avec un composant Link
```jsx
import Icon from './Icon';
import Link from '../../interface/Link';

<Link
    title="title"
    icon="ri-twitter-line"
    iconPosition="left"
    >this is a link
</Link>
```




