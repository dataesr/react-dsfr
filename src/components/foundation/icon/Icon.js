import React, { cloneElement, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './icons.css';

/**
 *
 * @visibleName Icon
 */
const Icon = ({
  size, name, as, children, className, iconPosition, title, verticalAlign, color,
}) => {
  const iconRef = useRef();
  const HTMLTag = `${as}`;
  const _className = classNames(
    `ri-${size}`,
    {
      [`icon-${iconPosition}`]: iconPosition !== 'center',
      'ds-fr--v-middle': verticalAlign,
    },
    name,
    className,
  );
  const i = title ? (
    <HTMLTag
      ref={iconRef}
      className={_className}
      title={title}
    />
  ) : (
    <HTMLTag
      ref={iconRef}
      className={_className}
    />
  );
  const newChildren = (
    <>
      {iconPosition === 'right' ? children?.props?.children : i}
      {iconPosition === 'right' ? i : children?.props?.children}
    </>
  );

  useEffect(() => {
    if (color && iconRef.current) {
      iconRef.current.style.color = `${color}`;
    }
  }, [color]);

  return (
    <>
      {children ? cloneElement(children, {
        ...children.props,
        className: classNames({
          [`${children.props.className}`]: children.props.className,
        }),
        children: newChildren,
      }) : i}
    </>
  );
};

Icon.defaultProps = {
  size: 'sm',
  color: '',
  as: 'span',
  className: '',
  verticalAlign: false,
  iconPosition: 'left',
  children: null,
  title: '',
};

Icon.propTypes = {
  size: PropTypes.oneOf(['fw', 'xxs', 'xs', 'sm', '1x', 'lg', 'lg', 'xl', '2x', '3x', '10x']),
  name: PropTypes.string.isRequired,
  verticalAlign: PropTypes.bool,
  className: PropTypes.string,
  iconPosition: PropTypes.oneOf(['left', 'right', 'center']),
  as: PropTypes.oneOf(['span', 'i', 'div', 'p']),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
  ]),
  title: PropTypes.string,
  color: PropTypes.string,
};

export default Icon;
