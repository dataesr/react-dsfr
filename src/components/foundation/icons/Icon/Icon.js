import PropTypes from 'prop-types';

/**
 * Design system Icon
 *
 * @visibleName Icon
 */
const Icon = ({ icon, size }) => {
  const iconSize = (size !== 'md') ? ` rf-fi--${size}` : '';
  return <span className={`rf-fi-${icon}${iconSize}`} />;
};

Icon.propTypes = {
  /**
  * Icon to use
  */
  icon: PropTypes.string.isRequired,
  /**
  * Icon size
  */
  size: PropTypes.oneOf(['md', 'lg', 'xl']),
};
Icon.defaultProps = {
  size: 'md',
};

export default Icon;
