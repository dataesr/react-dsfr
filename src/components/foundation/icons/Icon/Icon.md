Sélection proposée d’icônes

Des icônes doivent être utilisées pour attirer l'attention sur les actions, les ensembles de contenus importants ou les zones clés, il faut éviter d’en utiliser trop sur une même page pour ne pas créer de confusion.

Un concept doit être représenté par la même icône sur l'ensemble du site et de l'écosystème numérique de l'État.

```
<Icon icon="alert-fill" size="xl" />
```
