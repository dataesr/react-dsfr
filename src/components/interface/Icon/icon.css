/* ------------------------------------ *\
  ICONS
\* ------------------------------------ */

[class^="ri-"]:before,
[class*=" ri-"]:before {
    font-family: 'remixicon', serif !important;
    font-style: normal;
    font-weight: normal !important;
    font-variant: normal;
    text-transform: none;
}

.icon-v-align span,
.icon-v-align i {
    vertical-align: text-top;
}

[class^="ri-"].icon-left {
    margin-right: .5rem;
}

[class^="ri-"].icon-right {
    margin-left: .5rem;
}

.ri-2x.icon-right {
    margin-right: -.375rem;
    margin-left: .5rem;
}