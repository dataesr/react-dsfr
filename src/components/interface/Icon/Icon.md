## Icons - Icônes

Tous les icônes de la librairie [remixicon](https://remixicon.com/) sont utilisables.

### Usage simple
```jsx
import Icon from './Icon';

<Icon name="ri-bubble-chart-line">
    <p>Lorem ipsum</p>
</Icon>
```

```jsx
<Icon name="ri-flag-line" iconPosition="right" size="lg">
    <p>Lorem ipsum</p>
</Icon>
```

### Usage avec un composant Button
```jsx
import Icon from './Icon';
import Button from './../Button/Button';

<Button
    title="title"
    icon="ri-calendar-todo-fill"
    iconPosition="right">click
</Button>
```


