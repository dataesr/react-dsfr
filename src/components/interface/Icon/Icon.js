import React, { cloneElement } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import 'remixicon/fonts/remixicon.css';
import './icon.css';

/**
 *
 * @visibleName Icon
 */
const Icon = ({
  size, name, as, children, className, iconPosition,
}) => {
  const Tag = `${as}`;
  const _className = classNames(
    `ri-${size}`,
    { [`icon-${iconPosition}`]: iconPosition },
    name,
    className,
  );
  const i = <Tag className={_className} />;
  const newChildren = (
    <>
      {iconPosition === 'right' ? children.props?.children : i}
      {iconPosition === 'right' ? i : children.props?.children}
    </>
  );

  return (
    <>
      {cloneElement(children, {
        ...children.props,
        className: classNames({
          [`${children.props.className}`]: children.props.className,
        }),
        children: newChildren,
      })}
    </>
  );
};

Icon.defaultProps = {
  size: 'sm',
  as: 'span',
  className: '',
  iconPosition: 'left',
};

Icon.propTypes = {
  size: PropTypes.oneOf(['fw', 'xxs', 'xs', 'sm', '1x', 'lg', 'lg', 'xl', '2x', '3x', '10x']),
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  iconPosition: PropTypes.oneOf(['left', 'right']),
  as: PropTypes.oneOf(['span', 'i']),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
  ]).isRequired,
};

export default Icon;
