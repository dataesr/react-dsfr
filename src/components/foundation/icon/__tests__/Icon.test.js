import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import React from 'react';
import Icon from '../Icon';

const initialProps = {
  name: 'name',
  iconPosition: 'right',
  title: 'title',
  verticalAlign: true,
  as: 'span',
  size: 'sm',
  className: 'test',
  children: 'Icon text',
};

describe('<Icon />', () => {
  it('renders correctly', () => {
    const component = renderer
      .create(<Icon {...initialProps}><p>{initialProps.children}</p></Icon>)
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
