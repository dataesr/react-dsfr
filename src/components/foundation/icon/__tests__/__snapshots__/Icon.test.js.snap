// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> renders correctly 1`] = `
<p
  className=""
>
  Icon text
  <span
    className="ri-sm icon-right ds-fr--v-middle name test"
    title="title"
  />
</p>
`;
